### ISSUE TITLE
Monotonic Constraints for Temporal Fusion Transformer

### ISSUE THREAD

**Issue by StatMixedML** [jdb78/pytorch-forecasting#931]
Dear forecasting community,

in its [documentation](https://pytorch-forecasting.readthedocs.io/en/latest/api/pytorch_forecasting.models.temporal_fusion_transformer.TemporalFusionTransformer.html), the Temporal Fusion Transformer has the option of specifying monotonic constraints:

> monotone_constaints (Dict[str, int]) 
Dictionary of monotonicity constraints for continuous decoder variables mapping position (e.g. "0" for first position) 
to constraint (-1 for negative and +1 for positive, larger numbers add more weight to the constraint vs. the loss but 
are usually not necessary). This constraint significantly slows down training. Defaults to {}.

Unfortunately, this docstring is somewhat short and not self-contained. So my questions would be

- is it safe to assume that the monotonic constraints restrict the learned relationship between feature_x and response y, similar to [monotonic constrains in XGBoost](https://xgboost.readthedocs.io/en/stable/tutorials/monotonic.html)
- say we have three numeric features, i.e., x1, x2, x3 and a response y. Given the above explanation, would a positive monotone constraint for x2 be passed to the model as follows `monotone_constaints = {"x1": 0, "x2": 1, "x3": 0}`

Appreciate any help on this @jdb78.  

Many thanks

**Comment by jdb78** [comment]
Yes, this is correct. The implementation effectively penalizes a gradient wrt the relevant feature in the wrong direction. This operation 1st takes time and 2nd is not supported by some CUDA optimizations, so we need to fall back on a slower implementation. One alternative would be to calcualte pseudo gradients using pertubations. Probably faster and almost as effective. But not implemented, yet.

**Comment by StatMixedML** [comment]
@jdb78 Thanks for the fast reply. I close the issue now.
